tests/Doctrine/Tests/ORM/Functional/OneToManyToOneOrphanRemovalTest.php

```diff
@@ -0,0 +1,178 @@
+        $user2 = new User('user2', 'User 2');
+        $user3 = new User('user3', 'User 3');
+        $user4 = new User('user4', 'User 4');
+
+        $review1 = new Review('review1', 'review 1', 4, $user1);
+        $review2 = new Review('review2', 'review 2', 5, $user2);
+        $review3 = new Review('review3', 'review 3', 2, $user3);
+        $review4 = new Review('review4', 'review 4', 1, $user4);
+
+        $reviews = new Reviews();
+        $reviews->add($review1);
+        $reviews->add($review2);
+        $reviews->add($review3);
+        $reviews->add($review4);
+
+        $business = new Business('business1', 'business One', $reviews);
+
+        $this->_em->persist($business);
+        $this->_em->flush();
+
+        $this->businessId = $business->getId();
+
+        $this->_em->clear();
+    }
+
+    public function testOrphanRemoval(): void
+    {
+        $businessProxy = $this->_em->getReference(Business::class, $this->businessId);
+
+        $this->_em->remove($businessProxy);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $query  = $this->_em->createQuery('SELECT b FROM Doctrine\Tests\Models\Rating\Business b');
+        $result = $query->getResult();
+
+        self::assertCount(0, $result, 'Business should be removed by EntityManager');
+
+        $query  = $this->_em->createQuery('SELECT r FROM Doctrine\Tests\Models\Rating\Review r');
+        $result = $query->getResult();
+
+        self::assertCount(0, $result, 'Review should be removed by orphanRemoval');
+
+        $query  = $this->_em->createQuery('SELECT u FROM Doctrine\Tests\Models\Rating\User u');
+        $result = $query->getResult();
+
+        self::assertCount(0, $result, 'User should be removed by orphanRemoval');
+    }
+
+    public function testOrphanRemovalRemoveFromCollection(): void
+    {
+        $business = $this->_em->find(Business::class, $this->businessId);
+
+        $reviews = $business->getReviews();
+        $reviews->remove(1);
+        $business->setReviews($reviews);
+
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $query  = $this->_em->createQuery('SELECT r FROM Doctrine\Tests\Models\Rating\Review r');
+        $result = $query->getResult();
+
+        self::assertCount(3, $result, 'Review should be removed by orphanRemoval');
+
+        $query  = $this->_em->createQuery('SELECT u FROM Doctrine\Tests\Models\Rating\User u');
+        $result = $query->getResult();
+
+        self::assertCount(3, $result, 'User should be removed by orphanRemoval');
+    }
+
+    public function testOrphanRemovalRemoveWithExternallyBuiltArrayCollection(): void
+    {
+        $business = $this->_em->find(Business::class, $this->businessId);
+
+        $newReviews    = new Reviews();
+        $newReviewUser = new User('newUser', 'New User');
+        $newReviews->add(new Review('newReview1', 'new review 1', 4, $newReviewUser));
+
+        $business->setReviews($newReviews);
```

Comment:
@stof  this is an example of how my app might like to update an entity by just replacing the whole collection. In my experience with other solutions if the objects passed have the same IDs they are reconciled and or removed according to the new object. However in this use case, for instance here, the result is that the previous persisted collection had 4 items. I set a new collection with one item. I would expect the old items to be replaced/deleted and the new one to be saved (including the many to one relation) 
However the result is that they new review is added and the total becomes 5 .

Reply:
Well, as I said on Slack, if you replace the PersistentCollection doing the tracking, you cannot expect the change tracking to work.

Reply:
I am not replacing PersistentCollections but indeed just using a Collection provided by an external request, such as an update to a DTO, and feeding this DTO, that contains Collections, to the repository.

It is clear to me now how this whole thing work, ie you need to perform reconciliation and cannot use PPO, ie DTOs, directly with Doctrine save unless you are indeed creating a new record for sure.

Ideally this should be more obvious from documentation, usually showing rather simple use cases but also it would really be great if a reconciliation mechanism would be set in place. The latter is probably complex such as a serializer/mapping logic but it would really streamline interaction with repositories and make them very slim.

About documentation the only thing i see close to this matter is in this section
https://www.doctrine-project.org/projects/doctrine-orm/en/current/tutorials/getting-started.html#adding-behavior-to-entities

Here there is a User -> updateFromProfile function supposed to be responsible to do such reconciliation but again it is not clear. The only way I could see this working is by adding logic or a processor within the repository because with complex relations you need to continuously dig down the structure and take care of checking if an item is in a collection, update it if it is, remove it if removed in the DTO and take care of all its sub relations so the entity manager is needed continuously.


I will close the PR for now as it does not look like it is of interest anyway and it is rather complex for me to even design a possible solutinon

Reply:
your setter replaces the persistent collection with the external collection, instead of modifying the persistent collection that tracks changes done in the relation.